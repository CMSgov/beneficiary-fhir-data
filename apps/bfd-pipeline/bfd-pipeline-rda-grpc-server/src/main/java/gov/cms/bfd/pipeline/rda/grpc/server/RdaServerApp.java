package gov.cms.bfd.pipeline.rda.grpc.server;

import gov.cms.mpsm.rda.v1.FissClaimChange;
import gov.cms.mpsm.rda.v1.McsClaimChange;
import io.grpc.Server;
import java.io.File;
import java.io.IOException;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A stand-alone mock RDA API (version 0.2 MVP) server implementation. The server is intended for
 * testing purposes only and will not be used in production. Data served is specified on the command
 * line and comes from either random data or from a NDJSON file. The server always starts on port
 * 5003.
 */
public class RdaServerApp {
  private static final Logger LOGGER = LoggerFactory.getLogger(RdaServerApp.class);

  /**
   * Starts a RDA API server listening on localhost at a specific port. Configuration is controlled
   * by command line arguments. Each argument specifies one setting. Valid arguments are:
   *
   * <ul>
   *   <li>maxToSend:number sets the maximum number of objects to send in random streams
   *   <li>port:number sets the port for the server to listen on (default is 5003)
   *   <li>seed:number creates a random source using the number as the PRNG seed value
   *   <li>random creates a random source using current time as the PRNG seed
   *   <li>fissFile:filename creates a source that returns FissClaims contained in an NDJSON file
   *   <li>mcsFile:filename creates a source that returns McsClaims contained in an NDJSON file
   * </ul>
   */
  public static void main(String[] args) throws Exception {
    final Config config = new Config(args);
    LOGGER.info("Starting server on port {}.", config.getPort());
    Server server =
        RdaServer.startLocal(config.getPort(), config::createFissClaims, config::createMcsClaims);
    server.awaitTermination();
    LOGGER.info("server stopping.");
  }

  private static class Config {
    private final int port;
    private final long seed;
    private final int maxToSend;
    @Nullable private final File fissClaimFile;
    @Nullable private final File mcsClaimFile;

    private Config(String[] args) throws Exception {
      int port = 5003;
      long seed = System.currentTimeMillis();
      int maxToSend = 5_000;
      File fissClaimFile = null;
      File mcsClaimFile = null;
      for (String arg : args) {
        if (arg.startsWith("port:")) {
          port = Integer.parseInt(arg.substring(5));
        } else if (arg.startsWith("maxToSend:")) {
          maxToSend = Integer.parseInt(arg.substring(10));
        } else if (arg.startsWith("seed:")) {
          seed = Long.parseLong(arg.substring(5));
        } else if (arg.equals("random")) {
          seed = Long.parseLong(arg.substring(5));
        } else if (arg.startsWith("fissFile:")) {
          fissClaimFile = new File(arg.substring(9));
        } else if (arg.startsWith("mcsFile:")) {
          mcsClaimFile = new File(arg.substring(8));
        } else {
          throw new IOException("invalid argument: " + arg);
        }
      }
      this.port = port;
      this.seed = seed;
      this.maxToSend = maxToSend;
      this.fissClaimFile = fissClaimFile;
      this.mcsClaimFile = mcsClaimFile;
    }

    private int getPort() {
      return port;
    }

    private MessageSource<FissClaimChange> createFissClaims(long sequenceNumber) throws Exception {
      if (fissClaimFile != null) {
        LOGGER.info(
            "serving FissClaims using JsonClaimSource with data from file {}",
            fissClaimFile.getAbsolutePath());
        return new JsonMessageSource<>(fissClaimFile, JsonMessageSource::parseFissClaimChange);
      } else {
        LOGGER.info(
            "serving no more than {} FissClaims using RandomFissClaimSource with seed {}",
            maxToSend,
            seed);
        return new RandomFissClaimSource(seed, maxToSend).toClaimChanges().skip(sequenceNumber);
      }
    }

    private MessageSource<McsClaimChange> createMcsClaims(long sequenceNumber) throws Exception {
      if (mcsClaimFile != null) {
        LOGGER.info(
            "serving McsClaims using JsonClaimSource with data from file {}",
            mcsClaimFile.getAbsolutePath());
        return new JsonMessageSource<>(mcsClaimFile, JsonMessageSource::parseMcsClaimChange);
      } else {
        LOGGER.info(
            "serving no more than {} McsClaims using RandomMcsClaimSource with seed {}",
            maxToSend,
            seed);
        return new RandomMcsClaimSource(seed, maxToSend).toClaimChanges().skip(sequenceNumber);
      }
    }
  }
}
