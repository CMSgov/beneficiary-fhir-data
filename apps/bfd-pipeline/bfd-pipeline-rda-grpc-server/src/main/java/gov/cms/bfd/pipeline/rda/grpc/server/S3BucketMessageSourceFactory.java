package gov.cms.bfd.pipeline.rda.grpc.server;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import com.google.common.annotations.VisibleForTesting;
import gov.cms.bfd.pipeline.rda.grpc.RdaChange;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

/**
 * To save on the number of unnecessary calls to S3 to retrieve files and to allow the incremental
 * addition of files over time this implementation of {@link MessageSource.Factory} checks an S3
 * bucket for files matching a pattern that consists of a prefix, the range of sequence numbers in
 * the file, and a suffix. If no files matching this pattern are found but a file has a name equal
 * to the prefix followed by the suffix then that file is considered to hold all valid claims. When
 * one or more files match they are sorted by range and only those files containing sequence numbers
 * greater than or equal to the desired starting number are read. The matching files are then served
 * one after another in sequence number order.
 *
 * <p>NOTE: The files can have overlapping sequence numbers and/or sequence numbers that don't
 * correspond to the values in the object key. No effort is made to compensate for configuration
 * errors. As a consequence if the files contain data that differs from their file names or if they
 * have records out of order within the file the resulting stream of records might not be in
 * increasing order by sequence number.
 */
public class S3BucketMessageSourceFactory<T> implements MessageSource.Factory<T> {

  private final AmazonS3 s3Client;
  private final String bucketName;
  private final Function<String, MessageSource<T>> actualFactory;
  private final Function<T, Long> sequenceNumberGetter;
  private final Pattern matchPattern;

  public S3BucketMessageSourceFactory(
      AmazonS3 s3Client,
      String bucketName,
      String filePrefix,
      String fileSuffix,
      Function<String, MessageSource<T>> actualFactory,
      Function<T, Long> sequenceNumberGetter) {
    this.s3Client = s3Client;
    this.bucketName = bucketName;
    this.actualFactory = actualFactory;
    this.sequenceNumberGetter = sequenceNumberGetter;
    matchPattern =
        Pattern.compile(
            String.format("^%s(-(\\d+)-(\\d+))?\\.%s$", filePrefix, fileSuffix),
            Pattern.CASE_INSENSITIVE);
  }

  /**
   * Gets all available NDJSON files in the bucket that contain the specified sequence number and
   * creates a {@link MessageSource} that will return their messages.
   *
   * @param sequenceNumber minimum sequence number desired by the caller
   * @return a MessageSource pulling records from the bucket
   */
  @Override
  public MessageSource<T> apply(long sequenceNumber) throws Exception {
    List<FileEntry> entries = listFiles(sequenceNumber);
    return new MultiS3MessageSource(entries)
        .filter(record -> sequenceNumberGetter.apply(record) >= sequenceNumber);
  }

  /**
   * Searches the S3 bucket for objects whose key matches our regular expression. Only files whose
   * maximum sequence number is greater than or equal to our starting sequence number are retained.
   * The resulting list of entries is sorted by sequence number order.
   *
   * @param startingSequenceNumber smallest sequence number that the caller is interested in
   *     processing
   * @return a List of FileEntries containing the startingSequenceNumber and sorted by sequence
   *     number
   */
  @VisibleForTesting
  List<FileEntry> listFiles(long startingSequenceNumber) {
    List<FileEntry> entries = new ArrayList<>();
    List<S3ObjectSummary> summaries = s3Client.listObjects(bucketName).getObjectSummaries();
    for (S3ObjectSummary summary : summaries) {
      Matcher matcher = matchPattern.matcher(summary.getKey());
      if (matcher.matches()) {
        FileEntry entry;
        if (matcher.group(1) != null) {
          entry =
              new FileEntry(
                  summary.getKey(),
                  Long.parseLong(matcher.group(2)),
                  Long.parseLong(matcher.group(3)));
        } else {
          entry = new FileEntry(summary.getKey(), RdaChange.MIN_SEQUENCE_NUM, Long.MAX_VALUE);
        }
        if (entry.maxSequenceNumber >= startingSequenceNumber) {
          entries.add(entry);
        }
      }
    }
    entries.sort(FileEntry::compareTo);
    return entries;
  }

  /**
   * Compound {@link MessageSource} implementation that pulls records from a sequence of other
   * MessageSource. At any given time only one MessageSource is being consumed. MessageSources are
   * closed as they are completed.
   */
  private class MultiS3MessageSource implements MessageSource<T> {
    private final List<FileEntry> remaining;
    private MessageSource<T> current;

    /**
     * Constructs the source for the list of entries. Starts out with an empty source as current and
     * lets the first call to {@code next()} load the first entry lazily.
     *
     * @param remaining list of S3 entries to pull records from
     */
    private MultiS3MessageSource(List<FileEntry> remaining) {
      this.remaining = remaining;
      current = new EmptyMessageSource<>();
    }

    /**
     * Checks the current source for more records. If the current source has no more records it
     * closes that source and finds the next available source that does have a record. Sources are
     * closed along the way to ensure only the current source is open at any given time.
     *
     * @return true if next can be called successfully
     * @throws Exception if any operation fails
     */
    @Override
    public synchronized boolean hasNext() throws Exception {
      if (current.hasNext()) {
        return true;
      }
      while (remaining.size() > 0 && !current.hasNext()) {
        current.close();
        current = actualFactory.apply(remaining.remove(0).objectKey);
      }
      return current.hasNext();
    }

    @Override
    public synchronized T next() throws Exception {
      return current.next();
    }

    @Override
    public synchronized void close() throws Exception {
      current.close();
    }
  }

  /**
   * Immutable entry for S3 objects that match the key regex. Contains the min/max known sequence
   * number plus the key used to access the specific object in the bucket. Entries have a natural
   * order based on ascending min/max sequence numbers.
   */
  @VisibleForTesting
  @AllArgsConstructor
  @Getter
  @ToString
  @EqualsAndHashCode
  static class FileEntry implements Comparable<FileEntry> {
    private final String objectKey;
    private final long minSequenceNumber;
    private final long maxSequenceNumber;

    @Override
    public int compareTo(FileEntry o) {
      if (minSequenceNumber < o.minSequenceNumber) {
        return -1;
      } else if (minSequenceNumber > o.minSequenceNumber) {
        return 1;
      } else if (maxSequenceNumber < o.maxSequenceNumber) {
        return -1;
      } else if (maxSequenceNumber > o.maxSequenceNumber) {
        return 1;
      } else {
        return objectKey.compareTo(o.objectKey);
      }
    }
  }
}
