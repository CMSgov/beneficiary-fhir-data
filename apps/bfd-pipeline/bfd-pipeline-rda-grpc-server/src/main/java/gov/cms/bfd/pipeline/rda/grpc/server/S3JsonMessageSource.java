package gov.cms.bfd.pipeline.rda.grpc.server;

import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.S3ObjectInputStream;
import com.google.common.io.Closer;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

/**
 * Implementation of MessageSource that reads and serves NDJSON data from an object in an S3 bucket.
 * The object will be closed when the message source is closed. A JsonMessageSource.Parser object is
 * used to parse the NDJSON data into an appropriate object. Uses a Guava Closer to reliably close
 * all resources and abort the S3 stream if we are closed before all data has been consumed.
 *
 * <p>Design note: Mock servers only manage a small number of clients and generally only run for a
 * fixed period of time. To simplify the design the data os streamed to the client directly from S3.
 * This could create issues if the clients stall and never close their streams, etc. Since this
 * would only impact a test this isn't considered a problem worth solving through complicated logic
 * to download and cache files locally, etc.
 *
 * @param <T>
 */
public class S3JsonMessageSource<T> implements MessageSource<T> {
  private final Closer closer;
  private final S3ObjectInputStream s3InputStream;
  private final JsonMessageSource<T> jsonMessageSource;
  private boolean unfinished;

  public S3JsonMessageSource(S3Object s3Object, JsonMessageSource.Parser<T> parser) {
    closer = Closer.create();
    s3InputStream = s3Object.getObjectContent();
    // The S3 stream would normally download all the data from the object before closing.
    // Registering this method as a Closeable ensure that the abort method can be called on the
    // stream if we are closing before all the data has been read.
    closer.register(this::abortUnfinishedS3Stream);
    BufferedReader reader =
        new BufferedReader(new InputStreamReader(s3InputStream, StandardCharsets.UTF_8));
    // when the JsonMessageSource closes it will also close the S3 stream
    jsonMessageSource = closer.register(new JsonMessageSource<>(reader, parser));
    closer.register(s3Object);
    unfinished = true;
  }

  @Override
  public boolean hasNext() throws Exception {
    unfinished = jsonMessageSource.hasNext();
    return unfinished;
  }

  @Override
  public T next() throws Exception {
    return jsonMessageSource.next();
  }

  @Override
  public void close() throws IOException {
    closer.close();
  }

  /**
   * According to the S3 SDK docs, when an S3 stream is closed all the data remaining in the stream
   * will be downloaded before closing the stream. This method calls abort on the stream if we are
   * closing before we have retrieved all the data from the stream.
   */
  private void abortUnfinishedS3Stream() {
    if (unfinished) {
      s3InputStream.abort();
    }
  }
}
