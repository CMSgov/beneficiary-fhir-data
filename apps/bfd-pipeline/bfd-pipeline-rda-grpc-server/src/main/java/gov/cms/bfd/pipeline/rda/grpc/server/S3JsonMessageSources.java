package gov.cms.bfd.pipeline.rda.grpc.server;

import com.amazonaws.services.s3.AmazonS3;
import gov.cms.mpsm.rda.v1.FissClaimChange;
import gov.cms.mpsm.rda.v1.McsClaimChange;

/**
 * Uses an AmazonS3 client and a bucket name to simplify creation of {@link MessageSource}s that
 * read FISS or MCS claims from the bucket.
 */
public class S3JsonMessageSources {
  private final AmazonS3 s3Client;
  private final String bucketName;

  public S3JsonMessageSources(AmazonS3 s3Client, String bucketName) {
    this.s3Client = s3Client;
    this.bucketName = bucketName;
  }

  /**
   * Creates a {@link MessageSource} that reads {@link FissClaimChange} from an object in the
   * bucket.
   *
   * @param ndjsonObjectKey identifies the object containing our NDJSON data
   * @return a MessageSource that reads and parses the data
   */
  public MessageSource<FissClaimChange> readFissClaimChanges(String ndjsonObjectKey) {
    return createMessageSource(ndjsonObjectKey, JsonMessageSource::parseFissClaimChange);
  }

  /**
   * Creates a {@link MessageSource} that reads {@link McsClaimChange} from an object in the bucket.
   *
   * @param ndjsonObjectKey identifies the object containing our NDJSON data
   * @return a MessageSource that reads and parses the data
   */
  public MessageSource<McsClaimChange> readMcsClaimChanges(String ndjsonObjectKey) {
    return createMessageSource(ndjsonObjectKey, JsonMessageSource::parseMcsClaimChange);
  }

  private <T> MessageSource<T> createMessageSource(
      String ndjsonObjectKey, JsonMessageSource.Parser<T> parser) {
    return new S3JsonMessageSource<>(s3Client.getObject(bucketName, ndjsonObjectKey), parser);
  }
}
