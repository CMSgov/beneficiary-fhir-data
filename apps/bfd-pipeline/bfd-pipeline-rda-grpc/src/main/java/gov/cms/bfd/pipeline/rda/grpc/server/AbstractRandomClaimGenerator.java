package gov.cms.bfd.pipeline.rda.grpc.server;

import com.google.common.collect.ImmutableList;
import java.time.Clock;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

/**
 * The base class provides common functionality used by the FISS and MCS claim generators to create
 * random claim data.
 */
abstract class AbstractRandomClaimGenerator {
  private static final String ALPHA = "bcdfghjkmnpqrstvwxz";
  private static final String DIGIT = "1234567890";
  private static final String ALNUM = ALPHA + DIGIT;
  private static final int MAX_DAYS_AGO = 180;

  private final Random random;
  private final boolean optionalTrue;
  private final Clock clock;

  /**
   * Constructs an instance.
   *
   * @param seed numeric seed value for the PRNG
   * @param optionalTrue when true optionals will always be generated (used for tests)
   * @param clock Clock to generate current time/date values (needed for tests)
   */
  AbstractRandomClaimGenerator(long seed, boolean optionalTrue, Clock clock) {
    this.random = new Random(seed);
    this.optionalTrue = optionalTrue;
    this.clock = clock;
  }

  /**
   * RDA API enums always define a special value that the protobuf API intends as a special place
   * holder for garbled/invalid data. This field is named UNRECOGNIZED. This method is used to
   * accept all of the defined enum values, skip this special value, and return an immutable list of
   * valid enums to be included when generating a random enum value.
   *
   * @param values all possible values for an enum type
   * @param <T> the actual enum type
   * @return immutable list of valid enum values
   */
  protected static <T extends Enum<T>> List<T> enumValues(T[] values) {
    return Arrays.stream(values)
        .filter(v -> !v.name().equals("UNRECOGNIZED"))
        .collect(ImmutableList.toImmutableList());
  }

  protected int randomInt(int maxValue) {
    return maxValue == 0 ? 0 : random.nextInt(maxValue);
  }

  protected String randomDigit(int minLength, int maxLength) {
    return randomString(DIGIT, minLength, maxLength);
  }

  protected String randomLetter(int minLength, int maxLength) {
    return randomString(ALPHA, minLength, maxLength);
  }

  protected String randomAlphaNumeric(int minLength, int maxLength) {
    return randomString(ALNUM, minLength, maxLength);
  }

  /**
   * A random date string in YYYY-MM-DD format for a date no more than 100 days ago and no later
   * than today.
   *
   * @return random YYYY-MM-DD date string.
   */
  protected String randomDate() {
    final LocalDate date = LocalDate.now(clock).minusDays(random.nextInt(MAX_DAYS_AGO));
    return date.toString();
  }

  /**
   * Generates a random decimal string in the range 0.00 to 99999.99. The value will always have at
   * least one digit before the decimal point and exactly two digits after.
   *
   * @return random decimal string.
   */
  protected String randomAmount() {
    final int dollarDigits = 1 + random.nextInt(5);
    final StringBuilder amount = new StringBuilder(randomString(DIGIT, dollarDigits, dollarDigits));
    while (amount.length() > 1 && amount.charAt(0) == '0') {
      amount.setCharAt(0, randomChar(DIGIT));
    }
    amount.append(".").append(randomString(DIGIT, 2, 2));
    return amount.toString();
  }

  /**
   * Selects a random enum value from the list and returns it.
   *
   * @param values all possible values
   * @param <T> the enum type
   * @return one of the values selected at random
   */
  protected <T> T randomEnum(List<T> values) {
    return values.get(random.nextInt(values.size()));
  }

  /**
   * Triggers the action 50% of the time (or always if optionalTrue has been set).
   *
   * @param action action to trigger half the time
   */
  protected void optional(Runnable action) {
    if (optionalTrue || random.nextBoolean()) {
      action.run();
    }
  }

  /**
   * Used when one of several possible values should be generated. Usually used for enums. The
   * possibilities are triggered with equal probability.
   *
   * @param actions variadic list of possible actions to trigger
   */
  protected void oneOf(Runnable... actions) {
    final int index = random.nextInt(actions.length);
    actions[index].run();
  }

  /**
   * Used when an optional field can have one of several possible values. Usually used for enums. A
   * value will be triggered 50% of the time. When a value is triggered the possibilities are
   * selected with equal probability.
   *
   * @param actions variadic list of possible actions to trigger
   */
  protected void optionalOneOf(Runnable... actions) {
    optional(() -> oneOf(actions));
  }

  protected Clock getClock() {
    return clock;
  }

  private char randomChar(String characters) {
    return characters.charAt(random.nextInt(characters.length()));
  }

  private String randomString(String characters, int minLength, int maxLength) {
    final StringBuilder sb = new StringBuilder();
    final int len = minLength + random.nextInt(maxLength - minLength + 1);
    while (sb.length() < len) {
      sb.append(randomChar(characters));
    }
    return sb.toString();
  }
}
