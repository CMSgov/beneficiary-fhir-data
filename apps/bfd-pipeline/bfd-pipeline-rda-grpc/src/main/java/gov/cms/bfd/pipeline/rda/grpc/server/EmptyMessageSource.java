package gov.cms.bfd.pipeline.rda.grpc.server;

import java.util.NoSuchElementException;

/**
 * Trivial implementation of ClaimSource that returns no objects at all. Useful when only one type
 * of claim is needed for a given configuration of RdaService.
 *
 * @param <T>
 */
public class EmptyMessageSource<T> implements MessageSource<T> {

  /**
   * Factory method for use when creating RdaService instances that return no messages. Uses a
   * method rather than a static value so that types are inferred correctly by the compiler.
   *
   * @param <T> type of message being returned
   * @return a factory object
   */
  public static <T> MessageSource.Factory<T> factory() {
    return ignored -> new EmptyMessageSource<>();
  }

  @Override
  public boolean hasNext() throws Exception {
    return false;
  }

  @Override
  public T next() throws Exception {
    throw new NoSuchElementException();
  }

  @Override
  public void close() throws Exception {}
}
