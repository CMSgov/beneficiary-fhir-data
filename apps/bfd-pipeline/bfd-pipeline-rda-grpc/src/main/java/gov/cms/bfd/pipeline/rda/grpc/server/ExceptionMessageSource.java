package gov.cms.bfd.pipeline.rda.grpc.server;

import java.util.function.Supplier;

/**
 * Wrapper for a real MessageSource that throws an exception after some number of messages have been
 * delivered. Intended for use in integration tests that use RdaServer.
 *
 * @param <T> type of objects being delivered
 */
public class ExceptionMessageSource<T> implements MessageSource<T> {
  private final MessageSource<T> source;
  private final Supplier<Exception> exceptionFactory;
  private int remainingBeforeThrow;

  /**
   * Creates a new instance that delivers at most countBeforeThrow messages from source and then
   * throws an exception created by a call to exceptionFactory.
   *
   * @param source actual MessageSource
   * @param countBeforeThrow maximum number of messages to deliver before throwing
   * @param exceptionFactory factory to create the exception
   */
  public ExceptionMessageSource(
      MessageSource<T> source, int countBeforeThrow, Supplier<Exception> exceptionFactory) {
    this.source = source;
    this.exceptionFactory = exceptionFactory;
    remainingBeforeThrow = countBeforeThrow;
  }

  @Override
  public boolean hasNext() throws Exception {
    return source.hasNext();
  }

  @Override
  public T next() throws Exception {
    if (remainingBeforeThrow <= 0) {
      throw exceptionFactory.get();
    }
    remainingBeforeThrow -= 1;
    return source.next();
  }

  @Override
  public void close() throws Exception {
    source.close();
  }
}
