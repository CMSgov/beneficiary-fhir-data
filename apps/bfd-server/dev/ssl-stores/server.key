Bag Attributes
    friendlyName: server
    localKeyID: 54 69 6D 65 20 31 35 37 30 38 35 33 33 38 38 37 31 34 
Key Attributes: <No Attributes>
-----BEGIN PRIVATE KEY-----
MIIJQQIBADANBgkqhkiG9w0BAQEFAASCCSswggknAgEAAoICAQCjAlOHgB2QlUcb
O7jsZ/ow1qZCYLEa3JKv63n0Z+Rfp3/TZ0UPSQyhyFlZxqZB0/B05F/oBXW3RVJ5
wE6lpsRyoq2lprqaM3xSryrFljwCOVhO/SufRiBTkpB13TxWorcFDbYqhHd9sj0N
KxVcMuUdM4Ab+vPKiBe5B7f0isYdCHje3ERCLlCBLfrCYKvrK5hRVcdVQHvEmJ0w
3sHrt6yETaYYpobZLKBZaxPk15NA6DzZ9kKeP8q/ODP0H8LDTlKlBKUWF6ew1vUl
DRej3Lm7KDiL/8S1ji5nuBXnPgDNkeo2kWwe6DjblQqyujtXOyG0gEZ/Hli/aAU4
JBXeZLCJwi1yb8yJKuR7CzJ8ZKv/ZWDrK2BcvboWb87whlpZjrrmgiBkpFPDmEZs
L5t+Mkzjo4awQk4mialvZSo5oVyPnJszSHFqvBCZ0dJGRRA8ZQzy/EzKBrxbMxHe
zTSqcltIIG9IPeM3xNPGooB9OjdhtbiEpgF9autYWDCit8PJOnXZFfoz8dQrso3v
ILWn/Gpb+qrcIanNS1o/P2juEgxDPkLXosGJGPkPeqGjD5uSSy9isiyulcbetRpV
92FN40O0i/p9n5I5YH8oTze+kTY/cz9icoC0XA/ea+LXozmd+0Ld1/1FDYV5OIS4
nMTGiqLQFWJvAzISD2PQ4kVhhpzU1QIDAQABAoICACAkA0hCUJIDfQO3oNysG76p
dBq+ae079IhPS7qPCY56vk92mz1KT/1OPZ/f9UQhf9XseaSL/wvl1T9AVcF/wlhA
DzEMUNq6sBa1hx9yYd6mY+0fKgy7jQAtc+Jpfs4uQxvx+isj91OHY1ZHWmzVdCtv
ziC2OH+EmfjXtsO24R00YHDAyYyrCSCPalQNL5is/dl7qlP65Ox8kvHh2iaaDp//
l3Jg01Ha5Lr8y+1hPB2eLPFlBvmg5l8Ru9y0rnhH5y+2XL6gZ2YM3EUz8adTjCOE
6LFh+u05XZkByP/RT+M4lA+yWLk3XI6fMn7PDYhDln+6kyvZVfJXlj/djZpODCis
2OqLuodiiZzdfETzqskY7rBBd6nxORiXksONNfWwt/bVfUoLuWMoZD2tpDmEgUC7
/7wLflHcOrQe9SyLwujTn+akJM2oTpzIDZCO4k+7goOywhnFSvLyE2uJW9g3nsfI
/vTs63gr9vntbrXHK1JJYXuER8cgA1MzanaqW5uM45xaH+94PJFSahajl8sSRbTp
WMQyIIny4iD3lEDgrIA2FFbpMqAyoZLQamrYbAcWEmxfq6xT3fWIeOdsweg0IfBz
5vi61BKRknsV3W541XJxOMUK5/szNJh2i/4955IAZgz40u47rQKXy5ix/trfJatm
1CJiw9Z+bHXdSZr7l+i5AoIBAQDP3MT07n+omUfSN83bxH0zsxpWn8TDu5eKWbky
Agg7h36IzKnj/lfoANxJ2pVXCDYG5wL/jxYO/2nwVd2Th/P/rj4gOVjxOh6Y2ACl
BJ8WzkQX7QiMsJIhgxKSwJDMOa0XmYIoTSon4YgS3OSTjPGtDdLh09Ow7TnUEBpa
PtUdjsP8r+25kA/OCZhau1fKWJfDE7j5QsJIStEY0O99mEgQQq67UgixyvjV/mh4
PHfwE5QSRrEGxfliTWDYPo2GXyfbV15HAmsx1ZXsl+KTqDuka2nSR67u2mHy4kTH
Nx6NlRaC2J488DpJuPLAX9XeQTDlk8nt5YRhfaI9u5gEZpUHAoIBAQDIwmfj7Kzj
8W18PEmoVH6YET6K0ZBP1c6kkbjQ2RyxbfoShkHX65JnawsgFF70zzaFmAdgkyLQ
PLxmQzFxBBp5IdkkqfzZiEjtullZJM1U212Z954lR5Iht2sZOP32qesTepCOlcES
6zu9Y+YxwhA2jdTgXSRw641JGK+TPemaZzxy8ycc4uJruFecmn5Ga3zWIDM6Dz0G
3Mu+mbqRz4foIrDbhLYlSuemtpRHmCs7qH9Mfg+isKwjzMLtQ7K7Y6V/KQiRKorp
GrhgUOzI+jzj9luwK5ZNUSYxMo+7NMfHipEcpMT0kzhCohSZPkqJbHesY5nQCOSl
Kc/U2S9anIxDAoIBAADC3jXHadBEP28VpTtYzNyTm1UV4LKWzUdEbNkyCLWvGVeG
6FhdZhUDa8sZcdKPxSLhVbF55KrJZKnsppMTvGEDUlbr3teDTvlgUSBJlAuRVc6A
n7dkLiTRRhAjhGQ7Edc0SLW9EppyyPzIsk2YBoqLsmoBTJUpSTztkisib8Vb8HvM
aSJ01Jn8jwJGDVxAVs3wFI1VIUEChSGCIl5jYGAu3lTxR/bet2qRM/C8x/OpHrGy
cJOFxhQGFIS2Ne54Cr1S5JV9/MaB8bqcE8hWpWFEw25xtMJLKnCpQ3c3THVYRmzU
PkAa4RvEAZ+SukXGF1e/4s1sOb/hpeZz+WqKtaECggEAFZWBIeoiNGeSg9bffG+J
JMRF2xrozwyHqP3ZxupwwUq3piI1FgfHt8K0cYB/LQbSL87GjEWFaqn8/2md+S6d
wgQ3mk+EVcbnQFfWn1TrloH3MESFBjEJrwJPS2jrNQfTtEZrIElNdsE/rP7+kr7w
i0z4ZB7fehx5FjDHP6daZDVRNHKb8+kYusgY80dQ6DVKOKNBnq5gu0olxZo1itOr
IvQH27IK+ztwaGYczw23KDTXatvqxm0ENN0g3+11W9w8wM0ilKjkQmeTBxfsPTda
MJAeD3TANhPRbVyoQ1JLYFKvg0rbugfR9qgCWu9tiBTXUPtsYAd0vaUfKR0okC9g
jQKCAQBOzmUpPVCFwfC3rO8c7pOsqqj96CTDgScsd5t3xBhMkyMJpiVu7ud8xVhQ
4Kb+LGGiSBZR25S2hISE4y0rfpk/dUu9kHozTdVjNB61OsECuH5jiYjXvPM6Pf61
FJeLgSi1REsgWwzP04It/CCsXaeQmw4phCblYFYjA/9eGa/KfsCZHI9hDcBKNNlv
3LziF9gXKLEJlidVbZ9sZx6Ifklto0aWMbsu6p5f4Cyi+g/FjmwNPBpl2AHed9aT
bEdLJk+UI2qLzZFomrj9LGREEVGS2ImX0VqthUltVnfQ3pHHarbT0JPoCAfSXlE8
idNqxA47VZ0emxf17lCUOglT5SuK
-----END PRIVATE KEY-----
