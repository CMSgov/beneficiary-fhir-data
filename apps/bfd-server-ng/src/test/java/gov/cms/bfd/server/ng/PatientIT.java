package gov.cms.bfd.server.ng;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ca.uhn.fhir.context.FhirContext;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.context.annotation.Import;
import org.testcontainers.junit.jupiter.Testcontainers;

@ExtendWith({SnapshotExtension.class})
@Import(IntegrationTestConfiguration.class)
@Testcontainers
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class PatientIT {
  @LocalServerPort private int port;
  private Expect expect;

  @Test
  void patientRead() {
    FhirContext ctx = FhirContext.forR4();
    var client = ctx.newRestfulGenericClient("http://localhost:" + port + "/v3/fhir");
    var patient = client.read().resource("Patient").withId(1L).execute();

    expect.serializer("fhir+json").toMatchSnapshot(patient);
  }
}
