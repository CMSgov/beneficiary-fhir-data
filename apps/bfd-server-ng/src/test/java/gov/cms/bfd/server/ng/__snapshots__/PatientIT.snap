gov.cms.bfd.server.ng.PatientIT.patientRead=[
{
  "resourceType": "Patient",
  "id": "1",
  "meta": {
    "lastUpdated": "2024-01-01T00:00:00.000+00:00",
    "profile": [ "http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Patient|2.1.0", "http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient|6.1.0" ]
  },
  "extension": [ {
    "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-sex",
    "valueCode": "248152002"
  }, {
    "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race",
    "extension": [ {
      "url": "ombCategory",
      "valueCoding": {
        "system": "urn:oid:2.16.840.1.113883.6.238",
        "code": "2106-3",
        "display": "White"
      }
    }, {
      "url": "text",
      "valueCode": "White"
    } ]
  } ],
  "identifier": [ {
    "type": {
      "coding": [ {
        "system": "http://terminology.hl7.org/CodeSystem/v2-0203",
        "code": "MB"
      } ]
    },
    "system": "http://hl7.org/fhir/sid/us-mbi",
    "value": "1S000000000"
  } ],
  "name": [ {
    "family": "NORRIS",
    "given": [ "CHUCK", "R" ]
  } ],
  "gender": "female",
  "birthDate": "1960-01-01",
  "address": [ {
    "line": [ "100 MAIN ST" ],
    "city": "RICHMOND",
    "state": "VA",
    "postalCode": "22473"
  } ],
  "communication": [ {
    "language": {
      "coding": [ {
        "system": "urn:ietf:bcp:47",
        "code": "en"
      } ]
    }
  } ]
}
]